File: gluecodium/src/test/resources/smoke/field_constructors/output/swift/smoke/FieldConstructorsPartialDefaults.swift
//

//

import Foundation

public struct FieldConstructorsPartialDefaults {

    public var stringField: String

    public var intField: Int32

    public var boolField: Bool
    /// This is some field constructor with two parameters.
    /// It is very important.
    /// - Parameters
    ///   - intField: 
    ///   - stringField: 


    public init(intField: Int32, stringField: String) {
        self.intField = intField
        self.stringField = stringField
        self.boolField = true
    }


    public init(boolField: Bool, intField: Int32, stringField: String) {
        self.boolField = boolField
        self.intField = intField
        self.stringField = stringField
    }

    internal init(cHandle: _baseRef) {
        stringField = moveFromCType(smoke_FieldConstructorsPartialDefaults_stringField_get(cHandle))
        intField = moveFromCType(smoke_FieldConstructorsPartialDefaults_intField_get(cHandle))
        boolField = moveFromCType(smoke_FieldConstructorsPartialDefaults_boolField_get(cHandle))
    }
}



internal func copyFromCType(_ handle: _baseRef) -> FieldConstructorsPartialDefaults {
    return FieldConstructorsPartialDefaults(cHandle: handle)
}
internal func moveFromCType(_ handle: _baseRef) -> FieldConstructorsPartialDefaults {
    defer {
        smoke_FieldConstructorsPartialDefaults_release_handle(handle)
    }
    return copyFromCType(handle)
}

internal func copyToCType(_ swiftType: FieldConstructorsPartialDefaults) -> RefHolder {
    let c_stringField = moveToCType(swiftType.stringField)
    let c_intField = moveToCType(swiftType.intField)
    let c_boolField = moveToCType(swiftType.boolField)
    return RefHolder(smoke_FieldConstructorsPartialDefaults_create_handle(c_stringField.ref, c_intField.ref, c_boolField.ref))
}
internal func moveToCType(_ swiftType: FieldConstructorsPartialDefaults) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_FieldConstructorsPartialDefaults_release_handle)
}
internal func copyFromCType(_ handle: _baseRef) -> FieldConstructorsPartialDefaults? {
    guard handle != 0 else {
        return nil
    }
    let unwrappedHandle = smoke_FieldConstructorsPartialDefaults_unwrap_optional_handle(handle)
    return FieldConstructorsPartialDefaults(cHandle: unwrappedHandle) as FieldConstructorsPartialDefaults
}
internal func moveFromCType(_ handle: _baseRef) -> FieldConstructorsPartialDefaults? {
    defer {
        smoke_FieldConstructorsPartialDefaults_release_optional_handle(handle)
    }
    return copyFromCType(handle)
}

internal func copyToCType(_ swiftType: FieldConstructorsPartialDefaults?) -> RefHolder {
    guard let swiftType = swiftType else {
        return RefHolder(0)
    }
    let c_stringField = moveToCType(swiftType.stringField)
    let c_intField = moveToCType(swiftType.intField)
    let c_boolField = moveToCType(swiftType.boolField)
    return RefHolder(smoke_FieldConstructorsPartialDefaults_create_optional_handle(c_stringField.ref, c_intField.ref, c_boolField.ref))
}
internal func moveToCType(_ swiftType: FieldConstructorsPartialDefaults?) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_FieldConstructorsPartialDefaults_release_optional_handle)
}

File: gluecodium/src/test/resources/smoke/field_constructors/output/dart/lib/src/smoke/field_constructors_partial_defaults.dart


import 'dart:ffi';
import 'package:library/src/_library_context.dart' as __lib;
import 'package:library/src/builtin_types__conversion.dart';


class FieldConstructorsPartialDefaults {
  String stringField;

  int intField;

  bool boolField;

  /// This is some field constructor with two parameters.
  /// It is very important.
  /// [intField] 
  /// [stringField] 
  FieldConstructorsPartialDefaults.withTrue(this.intField, this.stringField)
      : boolField = true;
  FieldConstructorsPartialDefaults(this.boolField, this.intField, this.stringField);
}


// FieldConstructorsPartialDefaults "private" section, not exported.

final _smokeFieldconstructorspartialdefaultsCreateHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>, Int32, Uint8),
    Pointer<Void> Function(Pointer<Void>, int, int)
  >('library_smoke_FieldConstructorsPartialDefaults_create_handle'));
final _smokeFieldconstructorspartialdefaultsReleaseHandle = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_release_handle'));
final _smokeFieldconstructorspartialdefaultsGetFieldstringField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_get_field_stringField'));
final _smokeFieldconstructorspartialdefaultsGetFieldintField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Int32 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_get_field_intField'));
final _smokeFieldconstructorspartialdefaultsGetFieldboolField = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Uint8 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_get_field_boolField'));



Pointer<Void> smokeFieldconstructorspartialdefaultsToFfi(FieldConstructorsPartialDefaults value) {
  final _stringFieldHandle = stringToFfi(value.stringField);
  final _intFieldHandle = (value.intField);
  final _boolFieldHandle = booleanToFfi(value.boolField);
  final _result = _smokeFieldconstructorspartialdefaultsCreateHandle(_stringFieldHandle, _intFieldHandle, _boolFieldHandle);
  stringReleaseFfiHandle(_stringFieldHandle);
  
  booleanReleaseFfiHandle(_boolFieldHandle);
  return _result;
}

FieldConstructorsPartialDefaults smokeFieldconstructorspartialdefaultsFromFfi(Pointer<Void> handle) {
  final _stringFieldHandle = _smokeFieldconstructorspartialdefaultsGetFieldstringField(handle);
  final _intFieldHandle = _smokeFieldconstructorspartialdefaultsGetFieldintField(handle);
  final _boolFieldHandle = _smokeFieldconstructorspartialdefaultsGetFieldboolField(handle);
  try {
    return FieldConstructorsPartialDefaults(
      booleanFromFfi(_boolFieldHandle), 
      (_intFieldHandle), 
      stringFromFfi(_stringFieldHandle)
    );
  } finally {
    stringReleaseFfiHandle(_stringFieldHandle);
    
    booleanReleaseFfiHandle(_boolFieldHandle);
  }
}

void smokeFieldconstructorspartialdefaultsReleaseFfiHandle(Pointer<Void> handle) => _smokeFieldconstructorspartialdefaultsReleaseHandle(handle);

// Nullable FieldConstructorsPartialDefaults

final _smokeFieldconstructorspartialdefaultsCreateHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_create_handle_nullable'));
final _smokeFieldconstructorspartialdefaultsReleaseHandleNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_release_handle_nullable'));
final _smokeFieldconstructorspartialdefaultsGetValueNullable = __lib.catchArgumentError(() => __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('library_smoke_FieldConstructorsPartialDefaults_get_value_nullable'));

Pointer<Void> smokeFieldconstructorspartialdefaultsToFfiNullable(FieldConstructorsPartialDefaults? value) {
  if (value == null) return Pointer<Void>.fromAddress(0);
  final _handle = smokeFieldconstructorspartialdefaultsToFfi(value);
  final result = _smokeFieldconstructorspartialdefaultsCreateHandleNullable(_handle);
  smokeFieldconstructorspartialdefaultsReleaseFfiHandle(_handle);
  return result;
}

FieldConstructorsPartialDefaults? smokeFieldconstructorspartialdefaultsFromFfiNullable(Pointer<Void> handle) {
  if (handle.address == 0) return null;
  final _handle = _smokeFieldconstructorspartialdefaultsGetValueNullable(handle);
  final result = smokeFieldconstructorspartialdefaultsFromFfi(_handle);
  smokeFieldconstructorspartialdefaultsReleaseFfiHandle(_handle);
  return result;
}

void smokeFieldconstructorspartialdefaultsReleaseFfiHandleNullable(Pointer<Void> handle) =>
  _smokeFieldconstructorspartialdefaultsReleaseHandleNullable(handle);

// End of FieldConstructorsPartialDefaults "private" section.


